src/extras/reload.js

```diff
@@ -0,0 +1,69 @@
+(function (global) {
+  global.TRACING = true;
+
+  const systemJSPrototype = System.constructor.prototype;
+  const dependentsForModule = {};
+
+  systemJSPrototype.reload = function (id, parentUrl) {
+    const loader = this;
+
+    // An import first sets up a promise and resolves the module id, recreating
+    // the same behavior here avoids a race condition if a reload is called within
+    // that period
+    return Promise.resolve()
+      .then(function () {
+        return loader.resolve(id, parentUrl);
+      })
+      .then(function (id) {
+        const oldLoad = loader.getLoad(id);
+        if (!oldLoad) {
```

Comment:
This can be replaced by a `System.has` check on the resolved id.

Reply:
Actually, `System.has` will return false for modules that are _in progress_. Perhaps we should change `System.has` to indicate true for in progress loads, and use `System.get` as a way to see if the load is in success or failure state.

Reply:
Yes that would be a good change. 

The tricky part is that `System.import` doesn't put anything in the registry until after 1 microtask + `System.resolve`, so even when you're careful to check `System.has` you can still end up with a race condition. I'm not sure it's easy to solve this without changing `resolve` to be synchronous, I've worked around it in the reload function by also resolving the module id, so that the timing is the same.

Reply:
So any previous import would have already called `System.resolve` by this point. Thus our `System.resolve` would always happen after that previous `System.resolve` so I think there is effectively no gap then, as this path will always be slower than any other async path to get to the `id` check, such that any other `System.import` would have already defined the load.

The only way it seems that might break down is if `System.resolve` was slower for the first call than the second, but that seems a reasoable invariant to assume to me.